File: test/quick_canary_argv0.c
/* The C half of the argv[0]-independence canary (see quick_canary_argv0.pas).
   Compiled with NO -I on purpose: that is what makes lib/crtl resolve through
   ExeDir (the directory of ParamStr(0)), and those resolved module paths were
   being interned into the EMITTED string pool -- so the same compiler emitted
   './compiler/../lib/crtl/src/stdio.c' into the binary when run from the repo
   and nothing when run from a copy elsewhere. 218207 vs 218095 bytes.
   bug-a-the-compilers-output-depends-on-argv0 */
#include <stdio.h>
int main(void) { printf("c argv0 canary ok %d\n", 42); return 0; }

File: test/quick_canary_argv0.pas
{ The emitted binary must not depend on HOW THE COMPILER WAS INVOKED.

  A `uses` clause is the whole point of this file: unit resolution is where the
  compiler builds a path out of ExeDir (= the directory of ParamStr(0)), and
  that resolved path used to be interned into the EMITTED string pool. So the
  same compiler, compiling this same source, produced different bytes depending
  on whether it was run from inside the repo (where ExeDir + '../lib/rtl/'
  resolves) or from a copy elsewhere (where it falls back to the CWD-relative
  spelling). A program with no `uses` never noticed.
  bug-a-the-compilers-output-depends-on-argv0

  The assertion lives in the Makefile: compile this twice, from two different
  compiler paths, and cmp. Running it is the cheap half — a compile that emits
  nothing is not the property under test. }
program quick_canary_argv0;
uses sysutils;
begin
  writeln('argv0 canary ok ' + IntToStr(42));
end.
